examples/waveShared.cpp

```diff
@@ -0,0 +1,441 @@
+    fdBounds, fdBounds, [=](RAJA::Index_type ty, RAJA::Index_type tx) {
+
+      int id = tx + grid.nx * ty;
+      double x = grid.ox + tx * grid.dx;
+      double y = grid.ox + ty * grid.dx;
+      double myErr = std::abs(P[id] - waveSol(tf, x, y));
+
+      /*
+        tMax.max() is used to store the maximum value
+      */
+      tMax.max(myErr);
+    });
+
+  double lInfErr = tMax;
+  printf("Max Error = %lg, dx = %f \n", lInfErr, grid.dx);
+}
+
+
+/*
+ Function to set intial condition
+*/
+void setIC(double *P1, double *P2, double t0, double t1, grid_s grid)
+{
+
+  using myPolicy =
+    RAJA::NestedPolicy<RAJA::ExecList<RAJA::seq_exec, RAJA::seq_exec>>;
+  RAJA::RangeSegment fdBounds(0, grid.nx);
+  
+  RAJA::forallN<myPolicy>(
+    fdBounds, fdBounds, [=](RAJA::Index_type ty, RAJA::Index_type tx) {    
+
+      int id = tx + ty * grid.nx;
+      double x = grid.ox + tx * grid.dx;
+      double y = grid.ox + ty * grid.dx;
+      
+      P1[id] = waveSol(t0, x, y);
+      P2[id] = waveSol(t1, x, y);
+    });
+}
+
+
+//with shared memory
+template <typename T, typename fdNestedPolicy>
+void waveShared(T *P1, T *P2, RAJA::RangeSegment outerBounds, double ct, int nx)
+{
+
+#if CPU
+  //RAJA will give outer Id 
+ RAJA::forallN<fdNestedPolicy>(
+      outerBounds, outerBounds, [=] (RAJA::Index_type outerIdy, RAJA::Index_type outerIdx) {
+#else
+  //RAJA will give global id
+  RAJA::forallN<fdNestedPolicy>(
+      outerBounds, outerBounds, [=] __device__ (RAJA::Index_type gIdy, RAJA::Index_type gIdx) {
+        int outerIdx = blockIdx.x; //What I really want is the block id instead of thread of id... 
+        int outerIdy = blockIdx.y; //stick these in here...
+#endif
+      
+        /*
+          Coefficients for a fourth order stencil
+        */
+        double coeff[5] = {
+          -1.0 / 12.0, 4.0 / 3.0, -5.0 / 2.0, 4.0 / 3.0, -1.0 / 12.0};
+
+        //int lx, ly; //local index
+        int lx; int ly;
+        RAJA_shared double Lu[By+2*sr][By + 2*sr];       
+
+
+        //RAJA inner loop
+        RAJA_Par_Inner(ly, By) {
+          RAJA_Inner(lx, Bx) {
+
+            //Compute Index
+            const int tx = lx + Bx * outerIdx; 
+            const int ty = ly + By * outerIdy;
+            const int id = ty*nx+ tx; 
+            
+            //Index
+            const int nX1 = (tx - sr + nx)%nx; 
+            const int nY1 = (ty - sr + nx)%nx;
+            
+            const int nX2 = (tx + Bx - sr + nx)%nx;
+            const int nY2 = (ty + By - sr + nx)%nx;
+            
+            Lu[ly][lx] = P2[nY1*nx + nX1];
+            
+            if(lx < 2*sr){
+              Lu[ly][lx + Bx] = P2[nY1*nx + nX2];
+              
+              if(ly < 2*sr)
+                Lu[ly + By][lx + Bx] = P2[nY2*nx + nX2];
+            }
+            
+            if(ly < 2*sr)
+              Lu[ly + By][lx] = P2[nY2*nx + nX1];                      
+                        
+          } 
+        }
+        RAJA_sync;
```

Comment:
We've talked about having `RAJA::synchronize<POL>()` for a while for host-side synchronization. Maybe we could generalize something like that and use it here.